lib/ansible/plugins/connection/lxd.py

```diff
@@ -0,0 +1,110 @@
+# (c) 2016 Matt Clay <matt@mystile.com>
+#
+# This file is part of Ansible
+#
+# Ansible is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Ansible is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import os
+from distutils.spawn import find_executable
+from subprocess import call, Popen, PIPE
+
+import ansible.constants as C
+from ansible.errors import AnsibleError, AnsibleFileNotFound
+from ansible.plugins.connection import ConnectionBase
+from ansible.utils.unicode import to_str, to_bytes, to_unicode
+
+
+class Connection(ConnectionBase):
+    """ lxd based connections """
+
+    transport = "lxd"
+    has_pipelining = True
+
+    def __init__(self, play_context, new_stdin, *args, **kwargs):
+        super(Connection, self).__init__(play_context, new_stdin, *args, **kwargs)
+
+        self._host = self._play_context.remote_addr
+        self._lxc_cmd = find_executable("lxc")
+
+        if not self._lxc_cmd:
+            raise AnsibleError("lxc command not found in PATH")
+
+    def _connect(self):
+        """connect to lxd (nothing to do here) """
+        super(Connection, self)._connect()
+
+        self._connected = True
+
+    def exec_command(self, cmd, in_data=None, sudoable=True):
+        """ execute a command on the lxd host """
+        super(Connection, self).exec_command(cmd, in_data=in_data, sudoable=sudoable)
+
+        self._display.vvv(u"EXEC {0}".format(cmd), host=self._host)
+
+        executable = C.DEFAULT_EXECUTABLE.split()[0] if C.DEFAULT_EXECUTABLE else "/bin/sh"
```

Comment:
Looking over all the connection plugins, I'm not sure changing cmd to a list actually helps the situation. Every connection plugin passes cmd through a shell, either explicitly within the plugin or by way of delegating that responsibility to the remote host.

These plugins use subprocess with ```shell=true``` as the shell to run cmd:

- local

These plugins use ```C.DEFAULT_EXECUTABLE``` as the shell to run cmd:

- chroot
- docker
- jail
- libvirt_lxc
- lxd

These plugins let the remote system select the shell to run cmd:

- funcd
- paramiko_ssh
- ssh
- zone

So it seems all the plugins (including this one) are about about as consistent as they can be, although the local plugin could be updated to not use ```shell=true```.

Additionally, using ```C.DEFAULT_EXECUTABLE``` for the shell may not be the best choice, but if we're going to change that, we should do so for all of the plugins to keep things consistent.